src/workerd/api/headers.c++

```diff
@@ -138,54 +279,101 @@ jsg::Ref<Headers> Headers::clone(jsg::Lock& js) const {
 // Fill in the given HttpHeaders with these headers. Note that strings are inserted by
 // reference, so the output must be consumed immediately.
 void Headers::shallowCopyTo(kj::HttpHeaders& out) {
-  for (auto& entry: headers) {
-    for (auto& value: entry.second.values) {
-      out.addPtrPtr(entry.second.name, value);
+  for (kj::uint i = 1; i < commonHeaders.size(); i++) {
+    KJ_IF_SOME(header, commonHeaders[i]) {
+      KJ_IF_SOME(name, header->name) {
+        for (auto& value: header->values) {
+          out.addPtrPtr(name, value);
+        }
+      } else {
+        auto name = getCommonHeaderName(i);
+        for (auto& value: header->values) {
+          out.addPtrPtr(name, value);
+        }
+      }
     }
   }
-}
 
-bool Headers::hasLowerCase(kj::StringPtr name) {
-#ifdef KJ_DEBUG
-  for (auto c: name) {
-    KJ_DREQUIRE(!('A' <= c && c <= 'Z'));
+  // This is unfortunate... but we have to sort the uncommon headers by name
+  // before passing them off to kj::HttpHeaders...While the actual header order
+  // (for different header names) is not meaningful in HTTP, we have tests that
```

Comment:
Maybe we should just use a TreeMap instead of a HashMap, so they are maintained in order automatically?